magicctapipe/io/io.py

```diff
@@ -1419,3 +1420,132 @@ def resource_file(filename):
+    initial_time_offset : float
+        The initial time offset value for search the time offset
+
+    Returns
+    -------
+    t_magic_ave : float
+        The average the MAGIC timestamp
+    event_id_magic_ave : float
+        The average of the MAGIC event ID
+    time_offset_best : float
+        The best time offset found from the combination
+    n_coincident : float
+        The coincidence event number
+    """
+    if initial_time_offset is None:
+        time_offset_width1, time_offset_width2 = -4, 4
+        print(data_magic_["trigger_time"])
+        print(N_start)
+        time_select = (
+            data_lst_["trigger_time"].values
+            > (data_magic_["trigger_time"].values[N_start] + time_offset_width1)
+        ) & (
+            data_lst_["trigger_time"].values
+            < (data_magic_["trigger_time"].values[N_end] + time_offset_width2)
+        )
+        select = data_lst_["event_type"] == 32
+        if max(time_select) == True:
+            #N_start_ = N_start - len(data_magic_)
+            #N_end_ = N_end - len(data_lst_)
+            t_lst_all = data_lst_["trigger_time"].values[time_select & select]
+            print("size of LST time array ", len(t_lst_all))
+            t_magic_all = data_magic_[N_start:N_end]["trigger_time"].values
+
+            # compute time difference with all the combinations
+            array_1 = t_lst_all
+            array_2 = t_magic_all
+            comb_array = np.array(np.meshgrid(array_1, array_2)).reshape(2, -1)
+            time_offsets = comb_array[0] - comb_array[1]
+            print("size of combination array: ", np.shape(time_offsets))
+            time_window = 500e-9
+            b = np.array(
+                [np.sum(np.abs(time_offsets - T) < time_window) for T in time_offsets]
+            )
+            n_coincident = np.max(b)
+            time_offset_best = np.mean(time_offsets[b == np.max(b)])
+            std_b = np.std(b)
+            print("time offset: ", time_offset_best)
+            print("# of coincident events: ", n_coincident)
+            print("standard deviation: ", std_b)
+            return t_magic_all, N_start, time_offset_best, n_coincident
+        else:
+            print("No event was found")
+            return 0, 0, 0, 0
+
+    else:
+        EVENT_STEP = 15
+        N_ITR = len(data_magic_["trigger_time"].values[N_start:N_end]) // EVENT_STEP
+        t_magic_ave = np.zeros(N_ITR)
+        event_id_magic_ave = np.zeros(N_ITR)
+        time_offset_best = np.zeros(N_ITR)
+        n_coincident = np.zeros(N_ITR)
+
+        for n in range(N_ITR):
+            if n == 0:
+                time_offset_center = initial_time_offset
+            N_start_2 = n * EVENT_STEP
+            N_end_2 = (n + 1) * EVENT_STEP
+            time_offset_width1, time_offset_width2 = (
+                time_offset_center - 1.6e-6,
+                time_offset_center + 1.6e-6,
+            )
+            time_select = (
```

Comment:
what you do in each step is very similar to what you do at the beginning of the function with initial_time_offset None, maybe you can extract this into a separate function, or do the first case as 1 iteration case of multiple iterations?